stream: remove thenable support (#40773)
nodejs/node: doc/api/deprecations.md

[`readable.readableEnded`][] value on [`http.IncomingMessage`][] to get whether
it was an aborted or graceful destroy.

> 
> <!-- YAML

Reviewer: ```suggestion
<!-- YAML
```

> changes:
>   REPLACEME

Reviewer: ```suggestion
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/40773
    description: Documentation-only deprecation.
```

[`readable.readableEnded`][] value on [`http.IncomingMessage`][] to get whether
it was an aborted or graceful destroy.

> <!-- YAML

Reviewer 2: ```suggestion
### DEP0157: Thenable support in streams.

<!-- YAML
```

changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/40773
    description: Documentation-only deprecation.
-->

> Type: Runtime

Reviewer 2: Is it documentation-only or runtime? It needs to be consistent.

> Type: Documentation-only

Reviewer: This says `Documentation-only` but there are significant code changes....is this an oversight?

> Type: Documentation-only

Author: ```suggestion
Type: Runtime
```

> Streams no longer support thenables in implementation methods.

Reviewer 2: Can we guide users to another stream feature that would achieve the same as a thenable?
```suggestion
An undocumented feature of Node.js streams was to support thenables in implementation methods. This is now deprecated, instead use ???
```

[`readable.readableEnded`][] value on [`http.IncomingMessage`][] to get whether
it was an aborted or graceful destroy.

### DEP0157: Thenable support in streams

<!-- YAML
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/40773

>     description: Documentation-only deprecation.

Author: ```suggestion
    description: Runtime deprecation.
```

-->

Type: Runtime

> An undocumented feature of Node.js streams was to support thenables in implementation methods. This is now deprecated, instead use ???

Author: ```suggestion
An undocumented feature of Node.js streams was to support thenables in
implementation methods. This is now deprecated, instead use callbacks.
```

>   - version:
>       - v17.2.0
>       - v16.14.0
>   - version: REPLACEME
>     pr-url: https://github.com/nodejs/node/pull/40773
>     description: End-of-life.
>   - version: v17.2.0

Reviewer 2: ```suggestion
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/40773
    description: End-of-life.
  - version:
      - v17.2.0
      - v16.14.0
```

<!-- YAML
changes:

>   - version: REPLACEME
>     pr-url: https://github.com/nodejs/node/pull/40773

Reviewer 2: ```suggestion
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/40773
    description: End-of-life.
  - version: v17.2.0
    pr-url: https://github.com/nodejs/node/pull/40860
```

});
```

> ### DEP0XXX: Thenable support in streams

Reviewer 2: @ronag can you please rebase now that the doc-only deprecation has landed?
```suggestion
### DEP0157: Thenable support in streams
```

Reviewer 2: You need to merge this with https://github.com/nodejs/node/blob/79731df60843a47ed2c4c7b20da773d96157e4f0/doc/api/deprecations.md#L3022.

Author: I'm not sure I follow.

Reviewer 2: Instead of adding a new deprecation section, you need to re-use the existing one on line 3022.

Reviewer 2: Here's the diff, maybe that'd be clearer than me trying to find my words:

````diff
diff --git a/doc/api/deprecations.md b/doc/api/deprecations.md
index a543e4851c..5eb1cec12d 100644
--- a/doc/api/deprecations.md
+++ b/doc/api/deprecations.md
@@ -3023,12 +3023,15 @@ it was an aborted or graceful destroy.
 
 <!-- YAML
 changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/40773
+    description: End-of-life.
   - version: v17.2.0
     pr-url: https://github.com/nodejs/node/pull/40860
     description: Documentation-only deprecation.
 -->
 
-Type: Documentation-only
+Type: End-of-Life
 
 An undocumented feature of Node.js streams was to support thenables in
 implementation methods. This is now deprecated, use callbacks instead and avoid
@@ -3047,20 +3050,6 @@ const w = new Writable({
 });
 ```
 
-### DEP0XXX: Thenable support in streams
-
- <!-- YAML
- changes:
-   - version: REPLACEME
-     pr-url: https://github.com/nodejs/node/pull/40773
-     description: End-of-life.
- -->
-
-Type: End-of-Life
-
-An undocumented feature of Node.js streams was to support thenables in
-implementation methods. This is now deprecated, instead use callbacks.
-
 [Legacy URL API]: url.md#legacy-url-api
 [NIST SP 800-38D]: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
 [RFC 6066]: https://tools.ietf.org/html/rfc6066#section-3
````

Author: how about now?

Reviewer 2: You need to keep the history, see https://github.com/nodejs/node/pull/40773#discussion_r762551858.

<!-- YAML
changes:
  - version: REPLACEME
    pr-url: https://github.com/nodejs/node/pull/40773

>     description: Runtime deprecation.
> -->
> 
> Type: Runtime

Reviewer 2: This was discussed in TSC meeting today, no objections was raised to skip the runtime deprecation.
```suggestion
    description: End-of-life.
-->

Type: End-of-Life
```
